Add PSR-7 ServerRequestInterface to the PSR-7 Request

legionth opened the pull request:
This PR adds a [PSR-ServerRequestInterface](https://github.com/php-fig/fig-standards/blob/master/accepted/PSR-7-http-message.md#321-psrhttpmessageserverrequestinterface) to the request.

This PR adds the remote address to the server params(Fixes #153). Also cookies are parsed from the raw request and added to the request object.
Uploaded files are not handled in this PR, because this would be a much bigger task and needs an asynchronous approach(which would be too much to review).

The new class makes it possible to handle parsed request bodies (#105 for v0.8.0).

clue commented on src/ServerRequest.php:
```diff
@@ -0,0 +1,137 @@
+
+                if (!isset($cookies[$key])) {
+                    $result[$key] = $value;
+                }
```
Can you add some tests for multiple cookies with the same name and those using special chars?

clue replied:
Invalid reference?

clue commented on README.md:
```diff
@@ -136,12 +136,14 @@ connections and then processing each incoming HTTP request.
+This request object uses an implementation of the
 [PSR-7 RequestInterface](https://github.com/php-fig/fig-standards/blob/master/accepted/PSR-7-http-message.md#32-psrhttpmessagerequestinterface)
+and implements the
+[PSR-7 ServerRequestInterface](https://github.com/php-fig/fig-standards/blob/master/accepted/PSR-7-http-message.md#321-psrhttpmessageserverrequestinterface)
```
…implements the ServerRequestInterface which in turn extends the RequestInterface?

clue commented on README.md:
```diff
@@ -269,6 +273,20 @@ $http = new Server($socket, function (RequestInterface $request) {
 
+As mentioned before the request implements the `ServerRequestInterface`.
+The `Server` will add server-side parameters like the remote address or cookies
+this object.
```
…to this object.

clue commented on README.md:
```diff
@@ -269,6 +273,20 @@ $http = new Server($socket, function (RequestInterface $request) {
+    return new Response(
+        200,
+        array('Content-Type' => 'text/plain'),
+        "The remote: " . $request->getServerParams()['remote_address']
```
Where does this name come from, are there any other params? (needs some documentation)

clue commented on src/ServerRequest.php:
```diff
@@ -0,0 +1,137 @@
+
+        if (preg_match('(\,)', $cookie) >= 1) {
+            return false;
+        }
```
Should probably be moved up and could use some documentation? Looks like `strpos()`?

maciejmrozinski commented:
Just wanted to ask if we really need our own implementation of `Psr\Http\Message\ServerRequestInterface`? Cannot we use `Zend Diactoros` for example and extend it for our needs?

clue commented:
@maciejmrozinski I think you're raising a valid point here :+1: 

I'm perfectly fine with switching to any other PSR-7 implementation because we only expose its interfaces and everything else is basically just an implementation detail.

In this particular case, this PR only adds a very simple implementation of the `ServerRequestInterface` which extends an existing `Request` class and marks this new class as internal and not for outside usage. As such, this adds very little complexity afaict.

Switching to another implementation now would create a bigger changeset and would likely break compatibility with legacy PHP < 5.4 (which I'm not saying is a deal-breaker), but I suppose this is best left up to another PR? :+1:

maciejmrozinski commented:
@legionth OK, thx for the info, I totally agree :) Just wanted to know Your plan guys and possibilities here.

legionth commented:
Ping @clue. I think added all the changes you mentioned.

maciejmrozinski commented:
@legionth how about this https://github.com/reactphp/http/pull/160#pullrequestreview-30516086 ? Should it be changed or `RequestInterface` is correct here?

clue commented on README.md:
```diff
@@ -153,8 +155,29 @@ $http = new Server($socket, function (RequestInterface $request) {
+  Unix timestamp of the moment, the complete request header was received
+* `request_time_float`
+  Unix timestamp in microseconds of the moment, the complete request header wa
+  received
```
Where does this list come from? (ref please)

clue replied:
Example please?

legionth replied:
Added an reference to the text and an example.

clue commented on README.md:
```diff
@@ -153,8 +155,29 @@ $http = new Server($socket, function (RequestInterface $request) {
+  received
+
+The cookies of the request, will also be added to the request object by the `Server`.
+Use the `getCookies` method to receive these cookies.
```
Example please?

legionth replied:
Added example and more explanation for the cookies.

clue commented on README.md:
```diff
@@ -153,8 +155,29 @@ $http = new Server($socket, function (RequestInterface $request) {
+The `Server` will currently add several server-side parameters to the request like:
+
+* `server_address`
+  The current IP address of the server
```
What about the port?

clue commented on src/Server.php:
```diff
@@ -90,6 +106,7 @@
 class Server extends EventEmitter
 {
     private $callback;
+    private $serverAddress;
```
Should probably use `$conn->getLocalAddress()` instead in case this is a "listen all" address?

legionth replied:
Done. Restructured the code a bit an added an additional server parameter

legionth commented:
@clue Have look, I added some docuementation and test for the cookies and server params.

@maciejmrozinski Thank you! Changed the type hints. Have look

maciejmrozinski commented:
@legionth looks good to me

clue commented on README.md:
````diff
@@ -153,8 +155,41 @@ $http = new Server($socket, function (RequestInterface $request) {
+        $body
+    );
+});
+```
````
LGTM and should also be added to `examples/` :+1:

legionth replied:
Added to `examples/`

clue commented on README.md:
````diff
@@ -290,6 +325,76 @@ Allowed).
+        "Your cookie has been set."
+    );
+});
+```
````
LGTM, but should probably be merged with above example :+1:

legionth replied:
Merged the two examples into one and added to `examples/`

clue commented on tests/ServerRequestTest.php:
```diff
@@ -0,0 +1,164 @@
+    {
+        $cookieString = 'hello=world;hello=abc';
+        $cookies = ServerRequest::parseCookie($cookieString);
+        $this->assertEquals(array('hello' => 'world'), $cookies);
```
I'm not saying this is wrong, but this is certainly unexpected. Does this need explicit documentation or is this common behavior / does this follow a common implementation?

legionth replied:
Checked the `$_COOKIE` behavior of PHP. Changed the behavior a bit, have a look.

andig commented:
I'm working with @maciejmrozinski at porting php-pm to the next react/http version based on this PR. After only getting 400 responses I've dug into the `Server` code and found- using socket 0.5- the following issue:

  - `handleRequest` will throw a notice `port` not defined in https://github.com/reactphp/http/pull/160/files#diff-c611f5024b5b13f69f779505d1c6735eR279
  - because `$conn->getRemoteAddress` returns empty
  - because `stream_socket_get_name` returns `'' . "\0" . ''`. I've verified the stream is a valid resource

Any idea what might be wrong here?

andig commented:
**Update**

fixed for me by changing the socket `parseAddress()` to:

````
        if ($address == false || $address == '' . "\0" . '') {
            return null;
        }
````

Should I open an issue at socket?

andig commented:
pinging @clue I have the suspicion- though I can't find the root cause- that https://github.com/reactphp/socket/commit/a967ef1a8e95fccf77b6d4a6caf59805c88d9e57 might be causing (or not fixing) the problem I'm seeing. Apart from `false` php seems to return `string \0`, too.

clue commented:
@andig Thanks for spotting this and filing https://github.com/reactphp/socket/issues/92 which I think is the correct place to handle this :+1: 

---

@legionth This PR should probably be rebased now that #167 is in and we *should* probably add the equivalent of `$_SERVER['HTTPS'] = 'on'` here :+1:

legionth commented:
@clue I think this should cover your change requests. Added the `https` to the server params. Have a look.

legionth commented:
This PR adds a bit too much features and I can't cover them all. The fact is that this PR gets bigger and bigger, so I decided to open several smaller PRs.
